app/master/build.py

```diff
@@ -291,6 +295,15 @@ def _on_enter_error_state(self, event):
         self._error_message = getattr(event, 'error_msg', default_error_msg)
         self._logger.warning('Build {} failed: {}', self.build_id(), self._error_message)
 
+    def _on_leave_state(self, event):
+        start_time = self._state_machine.transition_timestamps.get(event.src)
```

Comment:
How do you feel about making this data internal to the BuildFSM object? It's already recording the timestamp values internally so it wouldn't be that weird for it also to be recording durations/deltas. Right now the timestamps are the "entered" time, so adding deltas would be equivalent to recording both "entered" and "exited" times.